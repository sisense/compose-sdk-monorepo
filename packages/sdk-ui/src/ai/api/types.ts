import { JaqlQueryPayload } from '@sisense/sdk-query-client';

import { ExpandedQueryModel } from '@/analytics-composer';
import { WidgetProps } from '@/props';

export interface ChatContext {
  title: string;
  live: boolean;
}

export interface ChatContextDetails {
  dashboardId?: string;
}
export interface TextMessage {
  role: 'assistant';
  content: string; // stringified json { answer: string }
  type: 'text';
}
export interface NlqMessage {
  role: 'assistant';
  content: string; // stringified json
  type: 'nlq';
}
interface RegularMessage {
  role: 'user' | 'assistant';
  content: string;
}
export type ChatMessage = RegularMessage | NlqMessage | TextMessage;

export interface Chat {
  chatId: string;
  contextId: string;
  contextTitle: string;
  contextDetails?: ChatContextDetails;
  expireAt: string;
  tenantId: string;
  userId: string;
  chatHistory: ChatMessage[];
}
export type ChatWithoutHistory = Omit<Chat, 'chatHistory'>;

export interface LlmConfig {
  provider: string;
  apiKey: string;
  baseUrl: string;
  model: string;
  version?: string;
}

export interface ChatRequest {
  text?: string;
  timezone?: string;
  options?: {
    enableFollowup?: boolean;
  };
}

interface TextResponse {
  chatHistory: ChatMessage[];
  data: {
    answer: string;
  };
  chatId: string;
  responseType: 'text';
}

interface ErrorResponse {
  data: {
    message: string;
    code: string;
    answer: string;
  };
  responseType: 'error';
}

/**
 * Response data for NLQ request
 */
export interface NlqResponseData extends ExpandedQueryModel {
  /** Detailed description of the response */
  detailedDescription: string;
  /** An array of suggested followup questions  */
  followupQuestions?: string[];
  /** NLQ prompt used in the request  */
  nlqPrompt: string;
  /** Suggested widget props*/
  widgetProps?: WidgetProps | undefined;
  /** The response message for the chat  */
  userMsg: string;
  /** @internal */
  clarification?: string;
}
export interface NlqResponse {
  data: NlqResponseData;
  timestamp: string;
  responseType: 'nlq';
}
export type ChatResponse = NlqResponse | TextResponse | ErrorResponse;

export interface QueryRecommendationConfig {
  numOfRecommendations: number;
  userPrompt?: string;
}

export type NlqResult = Omit<NlqResponseData, 'followupQuestions'>;

/** AI-generated recommended query you can run on your data model */
export interface QueryRecommendation extends Omit<NlqResponseData, 'followupQuestions'> {}
export type QueryRecommendationResponse = QueryRecommendation[];

export interface GetNlgInsightsRequest {
  jaql: Pick<JaqlQueryPayload, 'datasource' | 'metadata' | 'filterRelations'>;
  verbosity?: 'Low' | 'High';
}

/** @internal */
export interface GetNlgInsightsResponse {
  data?: {
    answer: string;
  };
  responseType: 'Text';
}

/** @internal */
export const NLQ_RESULT_CHART_TYPES = [
  'indicator',
  'column',
  'line',
  'table',
  'pie',
  'bar',
] as const;
/** NLQ result chart types @expandType */
export type NlqResultChartType = (typeof NLQ_RESULT_CHART_TYPES)[number];
export interface GetNlqResultRequest {
  text: string;
  timezone?: string;
  chartTypes: NlqResultChartType[];
}

export interface SendFeedbackRequest {
  type: string;
  data: object;
  sourceId: string;
  rating: -1 | 1;
}

/**
 * The chat mode to use for a chat session
 *
 * analyze: Enable business users to uncover data insights
 * develop: Enable developers to build queries and charts for embedded analytics
 */
export type ChatMode = 'analyze' | 'develop';
