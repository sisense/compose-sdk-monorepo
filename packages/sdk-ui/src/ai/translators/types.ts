import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '@/ai';
import { ChartType, TableType } from '@/types';
import { WidgetModel } from '@/models';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * All Chart Types including Table
 * Note: Table will be re-united with ChartType in a future MR
 *
 * @internal
 */
export type AllChartType = ChartType | TableType;

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart or table widget in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * Naming TBD
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps' | 'getTableWidgetProps'> & {
  chartType: AllChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl' | 'tableWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
