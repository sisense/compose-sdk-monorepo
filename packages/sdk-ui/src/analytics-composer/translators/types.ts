import { MetadataItem } from '@sisense/sdk-query-client';
import { AxesMapping, NlqResponseData } from '../../ai';
import { ChartType, WidgetModel } from '../../';

/**
 * Expanded Query Model that is based on NlqResponseData.
 * It contains expanded JAQL and chart recommendations returned from the chat response.
 *
 * @internal
 */
export type ExpandedQueryModel = Pick<
  NlqResponseData,
  'jaql' | 'chartRecommendations' | 'queryTitle'
>;

/**
 * Expanded Query Model
 *
 * Used as a default value for ExpandedQueryModel
 *
 * @internal
 */
export const EMPTY_EXPANDED_QUERY_MODEL: ExpandedQueryModel = {
  jaql: { datasource: { title: '' }, metadata: [] },
  queryTitle: '',
  chartRecommendations: {},
};

/**
 * Simple Chart Recommendations
 *
 * @internal
 */
export interface SimpleChartRecommendations {
  chartType: string;
  dataOptions: AxesMapping;
}

/**
 * Simple Query Model that is a simplified version of ExpandedQueryModel.
 *
 * It removes any JAQL syntax or information that already exists in the schema.
 * Also, any parameters that have a default do not need to be specified.
 *
 * Users manipulates this model via the query in YAML format.
 *
 * ModelTranslator is responsible for translating this query model to SimpleWidgetModel.
 *
 * @internal
 */
export interface SimpleQueryModel {
  /**
   * The data model or perspective title.
   */
  model: string;
  metadata: MetadataItem[];
  chart: SimpleChartRecommendations | {};
  /**
   * Title of the query
   */
  queryTitle: string;
}

/**
 * Empty Simple Query Model
 *
 * Used as a default value for SimpleQueryModel
 *
 * @internal
 */
export const EMPTY_SIMPLE_QUERY_MODEL: SimpleQueryModel = {
  model: '',
  metadata: [],
  chart: {},
  queryTitle: '',
};

/**
 * Chart Widget Model
 *
 * This model serves two purposes:
 *
 * (1) For rendering a chart widget (including table) in the Query Composer.
 * (2) For reverse engineered (translated) to CSDK code
 *
 * @internal
 */
export type ChartWidgetModel = Pick<WidgetModel, 'getChartWidgetProps'> & {
  chartType: ChartType;
  queryTitle: string;
};

/**
 * UI Framework
 *
 * @internal
 */
export type UiFramework = 'react' | 'vue' | 'angular';

/**
 * Code Template Key
 *
 * @internal
 */
export type CodeTemplateKey = 'baseChartTmpl' | 'chartTmpl' | 'chartWidgetTmpl';

/**
 * Code Templates
 *
 * @internal
 */
export type CodeTemplates = {
  [key in UiFramework]: { [key in CodeTemplateKey]: string };
};
