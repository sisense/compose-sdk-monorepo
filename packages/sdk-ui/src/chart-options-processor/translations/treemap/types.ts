import { Chart, PointLabelObject } from '@sisense/sisense-charts';

export type ParentValues = {
  direction: number;
  height: number;
  val: number;
  width: number;
  x: number;
  y: number;
};

export type LayoutPointResult = {
  x: number;
  y: number;
  width: number;
  height: number;
};

export type TreemapLayoutAlgorithmContext = {
  chart: Chart;
  squarified: (parent: ParentValues, children: PointLabelObject[]) => LayoutPointResult[];
};
