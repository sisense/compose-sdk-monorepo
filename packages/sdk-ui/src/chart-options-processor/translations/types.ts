import { AreamapChartDesignOptions } from '@/chart/restructured-charts/areamap-chart/types';
import { ChartType } from '../../types';
import {
  LineChartDesignOptions,
  AreaChartDesignOptions,
  BarChartDesignOptions,
  ColumnChartDesignOptions,
  CalendarHeatmapChartDesignOptions,
  PieChartDesignOptions,
  FunnelChartDesignOptions,
  IndicatorChartDesignOptions,
  PolarChartDesignOptions,
  ScatterChartDesignOptions,
  TreemapChartDesignOptions,
  BoxplotChartDesignOptions,
  ScattermapChartDesignOptions,
  AreaRangeChartDesignOptions,
  BaseDesignOptionsType,
} from './design-options';

/**
 * Style configuration for text elements in charts and UI components.
 * Defines the visual appearance and behavior of text rendering.
 */
export type TextStyle = {
  /**
   * Font family name for the text.
   * Examples: 'Arial', 'Helvetica', 'Times New Roman', 'sans-serif'
   */
  fontFamily?: string;
  /**
   * Font size in CSS units (px, em, rem, pt, etc.).
   * Examples: '12px', '1.2em', '16pt'
   */
  fontSize?: string;
  /**
   * Font weight for text rendering.
   * Examples: 'normal', 'bold', '100', '200', '300', '400', '500', '600', '700', '800', '900'
   */
  fontWeight?: string;
  /**
   * Font style for text rendering.
   * Examples: 'normal', 'italic', 'oblique'
   */
  fontStyle?: string;
  /**
   * Text color in any valid CSS color format.
   * Examples: '#FF0000', 'rgb(255, 0, 0)', 'red', 'rgba(255, 0, 0, 0.5)'
   */
  color?: string;
  /**
   * Text outline/stroke configuration.
   * Defines the outline color and width for text rendering.
   * Examples: '1px solid black', '2px #333'
   */
  textOutline?: string;
  /**
   * CSS pointer-events property for text interaction.
   * Controls how the text responds to mouse events.
   * Examples: 'auto', 'none', 'visible', 'visibleFill', 'visibleStroke'
   */
  pointerEvents?: string;
  /**
   * Text overflow handling behavior.
   * Controls how text is displayed when it exceeds its container.
   * Examples: 'clip', 'ellipsis', 'visible', 'hidden'
   */
  textOverflow?: string;
};

export const POLAR_CHART_TYPES = ['polar'] as const;
export type PolarChartType = (typeof POLAR_CHART_TYPES)[number];

export const CARTESIAN_CHART_TYPES = [
  'line',
  'area',
  'bar',
  'column',
  ...POLAR_CHART_TYPES,
] as const;
/** Cartesian family of chart types @expandType */
export type CartesianChartType = (typeof CARTESIAN_CHART_TYPES)[number];

export const CATEGORICAL_CHART_TYPES = ['pie', 'funnel', 'treemap', 'sunburst'] as const;
/** Categorical family of chart types  @expandType */
export type CategoricalChartType = (typeof CATEGORICAL_CHART_TYPES)[number];

export const SCATTER_CHART_TYPES = ['scatter'] as const;
/** Scatter chart types  @expandType */
export type ScatterChartType = (typeof SCATTER_CHART_TYPES)[number];

export const TABLE_TYPES = ['table'] as const;
/** Table chart types @expandType */
export type TableChartType = (typeof TABLE_TYPES)[number];

/** Table chart types @expandType */
export type TableType = TableChartType;

const INDICATOR_CHART_TYPES = ['indicator'] as const;
/** Indicator chart types @expandType */
export type IndicatorChartType = (typeof INDICATOR_CHART_TYPES)[number];

export const BOXPLOT_CHART_TYPES = ['boxplot'] as const;
/** Boxplot chart types  @expandType */
export type BoxplotChartType = (typeof BOXPLOT_CHART_TYPES)[number];

export const AREAMAP_CHART_TYPES = ['areamap'] as const;
/** Areamap chart types @expandType */
export type AreamapChartType = (typeof AREAMAP_CHART_TYPES)[number];

const IMAGE_CHART_TYPES = ['image'] as const;
export type ImageChartType = (typeof IMAGE_CHART_TYPES)[number];

export const SCATTERMAP_CHART_TYPES = ['scattermap'] as const;
/** Scattermap chart types  @expandType */
export type ScattermapChartType = (typeof SCATTERMAP_CHART_TYPES)[number];

export const CALENDAR_HEATMAP_CHART_TYPES = ['calendar-heatmap'] as const;
/**
 * Calendar heatmap chart types
 *
 * @expandType
 */
export type CalendarHeatmapChartType = (typeof CALENDAR_HEATMAP_CHART_TYPES)[number];

export const RANGE_CHART_TYPES = ['arearange'] as const;
/** AreaRange chart types  @expandType */
export type RangeChartType = (typeof RANGE_CHART_TYPES)[number];

// ChartDataType is the category of data structure for a group of charts,
// e.g. the ChartDataType of both "line" and "bar" charts is "cartesian",
// but a bubble/scatter chart would be a different data type.
export type ChartDataType =
  | 'cartesian'
  | 'categorical'
  | 'scatter'
  | 'table'
  | 'indicator'
  | 'areamap'
  | 'arearange';

/**
 * Design options for a specific chart type
 */
export type DesignOptions<SpecificChartType extends ChartType = ChartType> =
  SpecificChartType extends 'line'
    ? LineChartDesignOptions
    : SpecificChartType extends 'area'
    ? AreaChartDesignOptions
    : SpecificChartType extends 'bar'
    ? BarChartDesignOptions
    : SpecificChartType extends 'column'
    ? ColumnChartDesignOptions
    : SpecificChartType extends 'pie'
    ? PieChartDesignOptions
    : SpecificChartType extends 'funnel'
    ? FunnelChartDesignOptions
    : SpecificChartType extends 'polar'
    ? PolarChartDesignOptions
    : SpecificChartType extends 'indicator'
    ? IndicatorChartDesignOptions
    : SpecificChartType extends 'scatter'
    ? ScatterChartDesignOptions
    : SpecificChartType extends 'treemap'
    ? TreemapChartDesignOptions
    : SpecificChartType extends 'sunburst'
    ? TreemapChartDesignOptions
    : SpecificChartType extends 'boxplot'
    ? BoxplotChartDesignOptions
    : SpecificChartType extends 'areamap'
    ? AreamapChartDesignOptions
    : SpecificChartType extends 'scattermap'
    ? ScattermapChartDesignOptions
    : SpecificChartType extends 'calendar-heatmap'
    ? CalendarHeatmapChartDesignOptions
    : SpecificChartType extends 'arearange'
    ? AreaRangeChartDesignOptions
    : never;

/** A unique identifier for a series to be found in {@link ChartDataOptionsInternal} */
type SeriesId = string;

/**
 * Design options for a chart.
 * This includes global design options and specific design options per series.
 */
export type ChartDesignOptions<SpecificChartType extends ChartType = ChartType> =
  DesignOptions<SpecificChartType>;

/**
 * Design options, limited to the only series relevant options.
 *
 * @internal
 */
export type SeriesDesignOptions = Pick<BaseDesignOptionsType, 'marker'> &
  Pick<LineChartDesignOptions, 'line'>;
export type DesignPerSeries = { designPerSeries: Record<SeriesId, SeriesDesignOptions> };

export const isCartesian = (chartType: ChartType): chartType is CartesianChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return CARTESIAN_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isCategorical = (chartType: ChartType): chartType is CategoricalChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return CATEGORICAL_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isScatter = (chartType: ChartType): chartType is ScatterChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return SCATTER_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isScattermap = (chartType: ChartType): chartType is ScattermapChartType => {
  // Use .find instead of .includes for a more flexible type signature
  return SCATTERMAP_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isIndicator = (chartType: ChartType): chartType is 'indicator' => {
  return INDICATOR_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isPolar = (chartType: ChartType): chartType is PolarChartType => {
  return POLAR_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isTable = (chartType: ChartType | TableType): chartType is TableType => {
  return TABLE_TYPES.find((value) => value === chartType) !== undefined;
};

export const isBoxplot = (chartType: ChartType): chartType is BoxplotChartType => {
  return BOXPLOT_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isAreamap = (chartType: ChartType): chartType is AreamapChartType => {
  return AREAMAP_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isRange = (chartType: ChartType): chartType is RangeChartType => {
  return RANGE_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const isCalendarHeatmap = (chartType: ChartType): chartType is CalendarHeatmapChartType => {
  return CALENDAR_HEATMAP_CHART_TYPES.find((value) => value === chartType) !== undefined;
};

export const ALL_CHART_TYPES = [
  ...CARTESIAN_CHART_TYPES,
  ...CATEGORICAL_CHART_TYPES,
  ...SCATTER_CHART_TYPES,
  ...TABLE_TYPES,
  ...INDICATOR_CHART_TYPES,
  ...BOXPLOT_CHART_TYPES,
  ...AREAMAP_CHART_TYPES,
  ...IMAGE_CHART_TYPES,
  ...SCATTERMAP_CHART_TYPES,
  ...CALENDAR_HEATMAP_CHART_TYPES,
  ...RANGE_CHART_TYPES,
] as const;

export type DynamicChartType = (typeof ALL_CHART_TYPES)[number];

/*
export const isImage = (chartType: ChartType): chartType is ImageChartType => {
  return IMAGE_CHART_TYPES.find((value) => value === chartType) !== undefined;
};
*/
