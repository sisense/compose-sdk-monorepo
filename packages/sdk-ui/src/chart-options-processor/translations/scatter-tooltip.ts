import {
  InternalSeries,
  TooltipSettings,
  formatTooltipValue,
  formatTooltipXValue,
} from './tooltip-utils';
import { ScatterChartDataOptionsInternal } from '../../chart-data-options/types';
import { getDataOptionTitle } from '../../chart-data-options/utils';

export interface ScatterTooltipElements {
  x: string;
  y: string;
  breakByPoint: string;
  breakByColor: string;
  size: string;
}

export interface ScatterCustomPointOptions {
  maskedX: string;
  maskedY: string;
  maskedBreakByPoint?: string;
  maskedBreakByColor?: string;
  maskedSize?: string;
}

export const spanSegment = (value: number | string, color?: string): string =>
  value !== '' ? `<span style="fill:${color ?? 'currentColor'}">${value}</span>` : '';

const block = (element: string, title?: string): string => `${title || ''}<br />${element || ''}`;

const buildTooltip = (blocks: string[]): string =>
  blocks.reduce((tooltip, value) => tooltip + (value !== '<br />' ? `<br />${value}` : ''));

const buildSpans = (
  ctx: InternalSeries,
  dataOptions: ScatterChartDataOptionsInternal,
): ScatterTooltipElements => {
  const { maskedX, maskedY, maskedBreakByPoint, maskedSize, maskedBreakByColor } = ctx.point
    .custom as ScatterCustomPointOptions;

  const formatedX = formatTooltipXValue(dataOptions.x, ctx.point.x, maskedX);
  const formatedY = formatTooltipValue(dataOptions.y, ctx.point.y, maskedY);
  const formatedSize = formatTooltipValue(dataOptions.size, ctx.point.z, maskedSize || '');

  const x = spanSegment(formatedX, ctx.point.color);
  const y = spanSegment(formatedY, ctx.point.color);
  const breakByPoint = spanSegment(maskedBreakByPoint || '', ctx.point.color);
  const breakByColor = spanSegment(maskedBreakByColor || '', ctx.point.color);
  const size = spanSegment(formatedSize, ctx.point.color);

  return {
    x,
    y,
    breakByPoint,
    breakByColor,
    size,
  };
};

const buildBlocks = (
  spans: ScatterTooltipElements,
  dataOptions: ScatterChartDataOptionsInternal,
): ScatterTooltipElements => {
  const x = block(spans.x, dataOptions.x && getDataOptionTitle(dataOptions.x));
  const y = block(spans.y, dataOptions.y && getDataOptionTitle(dataOptions.y));
  const breakByPoint = block(
    spans.breakByPoint,
    dataOptions.breakByPoint && getDataOptionTitle(dataOptions.breakByPoint),
  );
  const breakByColor = block(
    spans.breakByColor,
    dataOptions.breakByColor && getDataOptionTitle(dataOptions.breakByColor),
  );
  const size = block(spans.size, dataOptions.size && getDataOptionTitle(dataOptions.size));

  return {
    x,
    y,
    breakByPoint,
    breakByColor,
    size,
  };
};

export const tooltipFormatter = (
  ctx: InternalSeries,
  dataOptions: ScatterChartDataOptionsInternal,
): string => {
  const spans = buildSpans(ctx, dataOptions);
  const { x, y, breakByPoint, breakByColor, size } = buildBlocks(spans, dataOptions);

  return buildTooltip([x, y, breakByPoint, breakByColor, size]);
};

export const getScatterTooltipSettings = (
  dataOptions: ScatterChartDataOptionsInternal,
): TooltipSettings => ({
  enabled: true,
  formatter(): string {
    const ctx: InternalSeries = this as InternalSeries;
    return tooltipFormatter(ctx, dataOptions);
  },
});
