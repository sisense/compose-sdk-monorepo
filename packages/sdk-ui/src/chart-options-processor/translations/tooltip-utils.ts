import { isNumber } from '@sisense/sdk-data';
import { Category, Value, isValue } from '../../chart-data-options/types.js';
import { applyFormat, defaultConfig } from './number-format-config.js';

export const isXValueNumeric = (dataOptionX: Value | Category | undefined) =>
  dataOptionX ? isValue(dataOptionX) || (dataOptionX.type && isNumber(dataOptionX.type)) : false;

export const formatTooltipValue = (
  dataOption: Value | Category | undefined,
  value: number | undefined,
  displayValue: string,
) => {
  if (!dataOption || value === undefined || isNaN(value)) return displayValue;

  return dataOption.numberFormatConfig
    ? applyFormat(dataOption.numberFormatConfig, value)
    : applyFormat(defaultConfig, value);
};

export const formatTooltipXValue = (
  dataOption: Value | Category | undefined,
  value: number | string | undefined,
  displayValue: string,
) => {
  const isNumeric = isXValueNumeric(dataOption);

  return isNumeric
    ? formatTooltipValue(dataOption, parseFloat(`${value}`), displayValue)
    : displayValue;
};

export type TooltipSettings = {
  enabled?: boolean;
  animation?: boolean;
  backgroundColor?: string;
  borderColor?: string;
  borderRadius?: number;
  borderWidth?: number;
  useHTML?: boolean;
  formatter?: () => string;
  style?: {
    fontFamily?: string;
  };
};

export type InternalSeries = {
  series: { name: string; color: string };
  x: string;
  y: number;
  point: {
    x?: number | string;
    y?: number;
    z?: number;
    name: string;
    color: string;
    custom?: {
      number1?: number;
      string1?: string;
      xDisplayValue?: string;
      xValue?: (number | string)[];
    };
  };
  percentage?: number;
};
