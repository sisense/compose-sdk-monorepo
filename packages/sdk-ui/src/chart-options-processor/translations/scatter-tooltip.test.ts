import { getScatterTooltipSettings, tooltipFormatter } from './scatter-tooltip';
import { ScatterChartDataOptionsInternal } from '../../chart-data-options/types';
import { InternalSeries } from './tooltip-utils';
import { TooltipFormatterContextObject } from '@sisense/sisense-charts';
import type { NumberFormatConfig } from '@/types';

describe('Scatter tooltip', () => {
  it('should be enabled', () => {
    const options = getScatterTooltipSettings({});
    expect(options.enabled).toBe(true);
  });

  it('When only x-axis', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: '0',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      x: {
        name: 'Total Costs',
      },
    } as ScatterChartDataOptionsInternal;

    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When x-axis and y-axis (call through options)', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
        },
      },
    } as unknown as TooltipFormatterContextObject;

    const dataOptions = {
      x: {
        name: 'Total Costs',
      },
      y: {
        name: 'Category',
      },
    } as ScatterChartDataOptionsInternal;

    const options = getScatterTooltipSettings(dataOptions);
    const tooltip = options.formatter?.call(tooltipContext);

    expect(tooltip).toMatchSnapshot();
  });

  it('When only y-axis', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '0',
          maskedY: 'Apple',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      y: {
        name: 'Category',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When x-axis, y-axis and break by / point', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
          maskedBreakByPoint: 'Sisense',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      x: {
        name: 'Total Costs',
      },
      y: {
        name: 'Category',
      },
      breakByPoint: {
        name: 'Brand',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When x-axis, y-axis, break by / point and size', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
          maskedBreakByPoint: 'Sisense',
          maskedSize: '124',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      x: {
        name: 'Total Costs',
      },
      y: {
        name: 'Category',
      },
      breakByPoint: {
        name: 'Brand',
      },
      size: {
        name: 'Total Costs',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When x-axis, y-axis, break by / point, color, and size', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
          maskedBreakByPoint: 'Sisense',
          maskedBreakByColor: 'Google',
          maskedSize: '124',
        },
      },
    } as InternalSeries;

    const dataOptions: ScatterChartDataOptionsInternal = {
      x: {
        name: 'Total Costs',
      },
      y: {
        name: 'Category',
      },
      breakByPoint: {
        name: 'Brand',
      },
      breakByColor: {
        name: 'Brand',
      },
      size: {
        name: 'Total Costs',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When only break by / point', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
          maskedBreakByPoint: 'Sisense',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      breakByPoint: {
        name: 'Brand',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When only break by / color', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
          maskedBreakByColor: 'Sisense',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      breakByColor: {
        name: 'Brand',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });

  it('When only size', () => {
    const tooltipContext = {
      series: { name: 'test', color: '#FFF' },
      x: '1',
      y: 1,
      point: {
        name: 'test',
        color: '#FFF',
        custom: {
          maskedX: '10M',
          maskedY: 'Apple',
          maskedSize: '10M',
        },
      },
    } as InternalSeries;

    const dataOptions = {
      size: {
        name: 'Total Costs',
      },
    } as ScatterChartDataOptionsInternal;
    const tooltip = tooltipFormatter(tooltipContext, dataOptions);

    expect(tooltip).toMatchSnapshot();
  });
});

it('Format numbers when x-axis, y-axis, break by / point, color, and size', () => {
  const format1: NumberFormatConfig = {
    name: 'Currency',
    symbol: '$',
    decimalScale: 1,
  };
  const format2: NumberFormatConfig = {
    name: 'Currency',
    symbol: '!',
    decimalScale: 1,
  };
  const format3: NumberFormatConfig = {
    name: 'Currency',
    symbol: '@',
    decimalScale: 1,
  };

  const tooltipContext = {
    series: { name: '3.14', color: '#FFF' },
    x: '22.10009',
    y: 2.234567,
    point: {
      name: '3.14',
      color: '#FFF',
      x: '22.10009',
      y: 2.234567,
      z: 12.345678,
      custom: {
        maskedX: '10M',
        maskedY: 'Apple',
        maskedBreakByPoint: '13.57976',
        maskedBreakByColor: '3.1456',
        maskedSize: '12.345678',
      },
    },
  } as InternalSeries;

  const dataOptions: ScatterChartDataOptionsInternal = {
    x: {
      name: 'x',
      type: 'number',
      numberFormatConfig: format1,
    },
    y: {
      name: 'y',
      numberFormatConfig: format2,
    },
    breakByPoint: {
      name: 'p',
      type: 'number',
      numberFormatConfig: format3,
    },
    breakByColor: {
      name: 'c',
      type: 'number',
      numberFormatConfig: format3,
    },
    size: {
      name: 'x',
      numberFormatConfig: format3,
    },
  } as ScatterChartDataOptionsInternal;
  const tooltip = tooltipFormatter(tooltipContext, dataOptions);

  expect(tooltip).toMatchSnapshot();
});
