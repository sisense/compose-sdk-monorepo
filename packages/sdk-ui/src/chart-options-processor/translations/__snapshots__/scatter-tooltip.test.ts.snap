// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Format numbers when x-axis, y-axis, break by / point, color, and size 1`] = `"x<br /><span style=\\"fill:#FFF\\">$22.1</span><br />y<br /><span style=\\"fill:#FFF\\">!2.2</span><br />p<br /><span style=\\"fill:#FFF\\">13.57976</span><br />c<br /><span style=\\"fill:#FFF\\">3.1456</span><br />x<br /><span style=\\"fill:#FFF\\">@12.3</span>"`;
